ext/ddtrace_profiling_native_extension/http_transport.c

```diff
@@ -0,0 +1,341 @@
+
+  if (exception_state) {           // An exception was raised
+    ddprof_ffi_Vec_tag_drop(tags); // clean up
+    rb_jump_tag(exception_state);  // "Re-raise" exception
+  }
+}
+
+// Note: This function handles a bunch of libddprof dynamically-allocated objects, so it MUST not use any Ruby APIs
+// which can raise exceptions, otherwise the objects will be leaked.
+static VALUE perform_export(
+  ddprof_ffi_NewProfileExporterV3Result valid_exporter_result, // Must be called with a valid exporter result
+  ddprof_ffi_Timespec start,
+  ddprof_ffi_Timespec finish,
+  ddprof_ffi_Slice_file slice_files,
+  ddprof_ffi_Vec_tag *additional_tags,
+  uint64_t timeout_milliseconds
+) {
+  ddprof_ffi_ProfileExporterV3 *exporter = valid_exporter_result.ok;
+  ddprof_ffi_CancellationToken *cancel_token = ddprof_ffi_CancellationToken_new();
+  ddprof_ffi_Request *request =
+    ddprof_ffi_ProfileExporterV3_build(exporter, start, finish, slice_files, additional_tags, timeout_milliseconds);
+
+  // We'll release the Global VM Lock while we're calling send, so that the Ruby VM can continue to work while this
+  // is pending
+  struct call_exporter_without_gvl_arguments args =
+    {.exporter = exporter, .request = request, .cancel_token = cancel_token, .send_ran = false};
+
+  // We use rb_thread_call_without_gvl2 instead of rb_thread_call_without_gvl as the gvl2 variant never raises any
+  // exceptions.
+  //
+  // (With rb_thread_call_without_gvl, if someone calls Thread#kill or something like it on the current thread,
+  // the exception will be raised without us being able to clean up dynamically-allocated stuff, which would leak.)
+  //
+  // Instead, we take care of our own exception checking, and delay the exception raising (`rb_jump_tag` call) until
+  // after we cleaned up any dynamically-allocated resources.
+  //
+  // We run rb_thread_call_without_gvl2 in a loop since an "interrupt" may cause it to return before even running
+  // our code. In such a case, we retry the call -- unless the interrupt was caused by an exception being pending,
+  // and in that case we also give up and break out of the loop.
+  int pending_exception = 0;
+
+  while (!args.send_ran && !pending_exception) {
+    rb_thread_call_without_gvl2(call_exporter_without_gvl, &args, interrupt_exporter_call, cancel_token);
+    if (!args.send_ran) pending_exception = check_if_pending_exception();
+  }
+
+  VALUE ruby_status;
+  VALUE ruby_result;
+
+  if (pending_exception) {
+    // We're in a weird situation that libddprof doesn't quite support. The ddprof_ffi_Request payload is dynamically
+    // allocated and needs to be freed, but libddprof doesn't have an API for dropping a request because it's kinda
+    // weird that a profiler builds a request and then says "oh, nevermind" and throws it away.
+    // We could add such an API, but I'm somewhat hesitant since this is a really bizarre corner case that looks quite
```

Comment:
I'll included it in the next version. I was just noticing independently we don't have such a thing.

Reply:
Uh, that's nice! Yeah `ddprof_ffi_Request_drop` works great for me, thanks!